events/dispatcher.go

```diff
@@ -0,0 +1,189 @@
+		}
+		go eventWorker.start()
+	}
+	log.Infof("events dispatcher started with %d workers", len(types.EventSources))
+}
+
+// Collect receives, check and push the events to the event queue
+func (d *Dispatcher) Collect(eventData proto.Message, sourceType SourceType) {
+	// Now, we take the delay, and the person's name, and make a WorkRequest out of them.
+	if sourceType < SourceType(0) || sourceType > SourceType(types.MaxUint8) {
+		log.Debugf("invalid source type: %d", sourceType)
+	}
+	// receive the event
+	event, err := d.wrap(eventData, sourceType)
+	if err != nil {
+		log.Debugf("cannot receive event: %w", err)
+		return
+	}
+
+	// add event to the database
+	if err := d.add(event); err != nil {
+		log.Debugf("cannot add event into database: %w", err)
+		return
+	}
+
+	log.Infof("event %+v added for processing", event)
+}
+
+// wrap checks and wraps an event into an Event struct
+func (d *Dispatcher) wrap(eventData proto.Message, sourceType SourceType) (*Event, error) {
+	event := &Event{}
+	switch sourceType {
+	case ScrutinizerOnprocessresult:
+		// check event data
+		if err := checkResults(eventData.(*models.ProcessResult)); err != nil {
+			return nil, fmt.Errorf("invalid event received: %w", err)
+		}
+	default:
+		return nil, fmt.Errorf("cannot receive event, invalid event origin")
+	}
+	event.Data = eventData
+	event.SourceType = sourceType
+	return event, nil
+}
+
+// DB
+
+// add adds an event to the events persistent storage.
+// An event will be stored as: []byte where byte[0] encodes the event type.
+// The other event bytes encode the event data itself
+func (d *Dispatcher) add(event *Event) error {
+	var eventDataBytes []byte
+	var err error
+	switch event.SourceType {
+	case ScrutinizerOnprocessresult:
+		switch t := event.Data.(type) {
+		case *models.ProcessResult:
+			eventDataBytes, err = proto.Marshal(t)
+			if err != nil {
+				return fmt.Errorf("cannot marshal event %+v data: %w", event, err)
+			}
+		default:
+			return fmt.Errorf("cannot add scrutinizer event to persisten storage, invalid event type")
+		}
+	default:
+		return fmt.Errorf(
+			"cannot add event %+v to persistent storage, unsupported event origin: %d",
+			event,
+			event.SourceType,
+		)
+	}
+	queueIdx := 0
+	switch est := event.SourceType; {
+	case est < types.ScrutinizerSourceTypeLimit:
+		// nothing to do, 0 is fine
+	case est >= types.ScrutinizerSourceTypeLimit && est < types.KeyKeeperSourceTypeLimit:
+		queueIdx = 1
+	case est >= types.KeyKeeperSourceTypeLimit && est < types.EthereumSourceTypeLimit:
+		queueIdx = 2
+	case est >= types.EthereumSourceTypeLimit && est < types.VochainSourceTypeLimit:
+		queueIdx = 3
+	case est >= types.VochainSourceTypeLimit && est < types.CensusSourceTypeLimit:
+		queueIdx = 4
+	default:
+		return fmt.Errorf("cannot add event, invalid queue index")
+	}
+	// byte[0]  -> event type
+	// byte[1:] -> event data
+	err = d.Storage[queueIdx].Put(
+		append(
+			[]byte(fmt.Sprintf("%x", event.SourceType)),
```

Comment:
I don't understand this. When you read the []byte messages from the disk queue, you use the first byte as the SourceType directly, without any decoding. But here, you encode the source type as hex.

In other words, you don't want to hex-encode here, I think. You can just do `append([]byte{event.SourceType}, eventDataBytes...)`